File: python/cote/LTCD/최소_높이_트리.py
from typing import List
from collections import defaultdict

def solve(edges: List[int], n: int) -> List[int] :
	if n <= 1 :
		return [0]

	graph = defaultdict(list)

	for i, j in edges :
		graph[i].append(j)
		graph[j].append(i)

	leaves = []

	for i in range(n) :
		if  len(graph[i]) == 1 :
			leaves.append(i)
	while n > 2 :
		n -= len(leaves)

		new_leaves = [] 
		for leaf in leaves :
			neighbor = graph[leaf].pop()
			graph[neighbor].remove(leaf)

			if len(graph[neighbor]) == 1:
				new_leaves.append(neighbor)
		leaves = new_leaves

	return leaves

if __name__ == "__main__" :
	with open("data/최소_높이_트리.txt") as f :

		testcases = int(f.readline())

		for _ in range(testcases) :
			n, m = map(int, f.readline().split(" "))
			edges = [ list(map(int, f.readline().split(" "))) for _ in range(m)]
			answer = list(map(int, f.readline().split(" ")))
			print(answer, solve(edges, n))

File: javascript/cote/LTCD/최소_높이_트리.mjs
import {readFile} from 'fs/promises'


const data = await readFile('data/최소_높이_트리.txt', {'encoding' : 'utf-8'})
	.then(lines => lines.split("\n").map(
		line => line.trim()))

let testCases = parseInt(data[0])
let cur = 1;

function solve(edges, n) {
	if (n <= 1)
		return [0]

	const graph = new Map()

	for (let i = 0; i < n; i++) {
		graph.set(i, [])
	}

	for (let [i, j] of edges) {
		graph.get(i).push(j)
		graph.get(j).push(i)
	}	

	let leaves = []

	for (let [i, node] of graph.entries()) {
		if(node.length == 1) {
			leaves.push(i)
		}
	}

	while (n > 2) {
		n -= leaves.length

		let new_leaves = []

		for (let leaf of leaves) {
			const parent = graph.get(leaf).pop()
			graph.set(parent, graph.get(parent).filter(i => i != leaf))

			if (graph.get(parent).length == 1){
				new_leaves.push(parent)
			}
		}
		leaves = new_leaves
	}

	return leaves
}

while (testCases > 0) {
	const [n, m] = data[cur++].split(" ").map(l => parseInt(l))
	const edges = []
	for (let i = 0; i < m; i++){
		edges.push(data[cur++].split(" ").map(l => parseInt(l)))
	}
	const answer = data[cur++].split(" ").map(l => parseInt(l))
	console.log(answer, solve(edges, n))
	testCases--;
}